### PR TITLE
KAFKA-3562: Handle topic deletion during a send

### PR DESCRIPTION
**Pull request by rajinisivaram**
Fix timing window in producer by holding onto cluster object while processing send requests so that changes to cluster during metadata refresh don't cause NPE if a topic is deleted.

### CODE REVIEW

**Comment by harshach** [comment]
+1

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 1]
```diff
@@ -455,7 +456,13 @@ private static int parseAcks(String acksString) {
                         " to class " + producerConfig.getClass(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG).getName() +
                         " specified in value.serializer");
             }
-            int partition = partition(record, serializedKey, serializedValue, metadata.fetch());
+            // If topic is deleted between the start of this method when metadata was obtained
+            // and now, fail the send request.
+            Cluster cluster = metadata.fetch();
```
Comment (ewencp) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Would it be better to just hold on to the `Cluster` object obtained at the start of this and use it throughout the call? I'm not sure how practical that is given the current code, but trying to handle inconsistencies in this data throughout the call seems like it'll be messy. For example, we accumulate messages and send based on topic partitions, does that also break if metadata changes? And do the issues highlighted here get even messier if the # of topic partitions change instead of the topic being entirely deleted?

Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: @ewencp Thank you for the review. I have updated the code to hold on to `Cluster` after a metadata update to avoid small timing windows like this issue in methods where metadata may have changed. For RecordAccumulator, it is better to get the latest metadata since a significant amount of time could have passed between accumulating and sending. Changes in the cluster are already handled properly in this case. Invalid partitions are already handled in the producer which covers changes to partitions, and since partitions cannot be deleted in Kafka, there is less scope for issues  compared to topic deletion.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 2]
```diff
@@ -138,7 +138,7 @@ public synchronized boolean updateRequested() {
     /**
      * Wait for metadata update until the current version is larger than the last version we know of
      */
-    public synchronized void awaitUpdate(final int lastVersion, final long maxWaitMs) throws InterruptedException {
+    public synchronized Cluster awaitUpdate(final int lastVersion, final long maxWaitMs) throws InterruptedException {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: Do we really need this change? If we immediately call `fetch` after `awaitUpdate`, we may get a subsequent version, but that shouldn't matter, right? The reason I ask is that it seems a bit arbitrary to return `Cluster` as opposed to other state held by `Metadata`.

Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java]: @ijuma Thank you for the review. You are right, this is an unnecessary change. No idea what I was thinking.

**Comment by ijuma** [comment]
Thanks @rajinisivaram, the change looks good to me now. Would it be possible to add a test?

**Comment by rajinisivaram** [comment]
@ijuma I had looked to see if there were unit tests around this area, but couldn't find any. Producer good path gets tested in integration tests, but it looks difficult to add a test for the failure in this JIRA without refactoring the producer implementation to make it easier to inject failures.

**Comment by ijuma** [comment]
@rajinisivaram, I thought we could use `EasyMock` and mock the producer and verify that `Metadata.fetch` only gets called the expected number of times. That would fail if people forget to reuse the returned `Cluster` instance. What do you think?

**Comment by rajinisivaram** [comment]
@ijuma Thank you, I imagine it would require `PowerMock` since metadata class is final and the field is private/final. I will take a look this evening.

**Comment by rajinisivaram** [comment]
@ijuma I have added unit tests.

**Comment by rajinisivaram** [comment]
@ijuma Can you take a look at the tests when you have time? Thanks.

**Comment by ijuma** [comment]
Thanks @rajinisivaram, sorry for the delay I've been travelling. I'll try to look into this soon.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 3]
```diff
@@ -123,4 +139,48 @@ public void testInvalidSocketReceiveBufferSize() throws Exception {
         config.put(ProducerConfig.RECEIVE_BUFFER_CONFIG, -2);
         new KafkaProducer<>(config, new ByteArraySerializer(), new ByteArraySerializer());
     }
+
+    @PrepareOnlyThisForTest(Metadata.class)
+    @Test
+    public void testMetadataFetch() throws Exception {
+        Properties props = new Properties();
+        props.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        KafkaProducer<String, String> producer = new KafkaProducer<String, String>(
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Nitpick: use diamond operator.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 4]
```diff
@@ -123,4 +139,48 @@ public void testInvalidSocketReceiveBufferSize() throws Exception {
...
+                props, new StringSerializer(), new StringSerializer());
+        Metadata metadata = PowerMock.createNiceMock(Metadata.class);
+        MemberModifier.field(KafkaProducer.class, "metadata").set(producer, metadata);
+
+        String topic = "topic";
+        ProducerRecord<String, String> record = new ProducerRecord<String, String>(topic, "value");
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Nitpick: use diamond operator.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java [hunk 5]
```diff
@@ -123,4 +139,48 @@ public void testInvalidSocketReceiveBufferSize() throws Exception {
...
+        PowerMock.replay(metadata);
+        producer.send(record, null);
+        PowerMock.verify(metadata);
+
+        // Expect exactly one fetch if topic metadata is available
+        PowerMock.reset(metadata);
+        EasyMock.expect(metadata.fetch()).andReturn(cluster).once();
+        PowerMock.replay(metadata);
+        producer.partitionsFor(topic);
+        PowerMock.verify(metadata);
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Do these checks actually work? I added an additional `metadata.fetch()` call in `waitOnMetadata` and the test still passed.

Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: @ijuma Thank you for the review. The code returns a valid cluster where specified, and a null cluster if more calls are made. For example, in lines 181-183 above, the first call to `metadata.fetch()` from `waitOnMetadata` returns a good cluster. If that cluster is used, `partitionsFor` works fine. But if you fetch again, you get a null cluster that will result in NullPointerException when you use the cluster. The original `partitionsFor`code was:

```
waitOnMetadata(topic, this.maxBlockTimeMs);
....
return this.metadata.fetch().partitionsForTopic(topic);
```

That fails the test because the second fetch returns null:

```
java.lang.NullPointerException
    at org.apache.kafka.clients.producer.KafkaProducer.partitionsFor(KafkaProducer.java:612)
    at org.apache.kafka.clients.producer.KafkaProducerTest.testMetadataFetch(KafkaProducerTest.java:183)
    ...
```

Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Thanks for the explanation. Would it be easy to change the code to throw an exception instead of returning null (in my test, I didn't actually use the return value and why it still worked fine)? If so, that is probably a little better. If not, then the current code is fine.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java (hunk 3 above)
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Done.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java (hunk 4 above)
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: Done.

File: clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java (hunk 5 above)
Comment (rajinisivaram) [inline clients/src/test/java/org/apache/kafka/clients/producer/KafkaProducerTest.java]: @ijuma Thank you, have added exception.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 6]
```diff
@@ -508,17 +511,16 @@ private static int parseAcks(String acksString) {
...
-     * @return The amount of time we waited in ms
+     * @param beginTimeMs Start time from which maxWaitMs applies
+     * @return The cluster containing topic metadata
      */
-    private long waitOnMetadata(String topic, long maxWaitMs) throws InterruptedException {
+    private Cluster waitOnMetadata(String topic, long maxWaitMs, long beginTimeMs) throws InterruptedException {
         // add topic to metadata topic list if it is not there already and reset expiry
         this.metadata.add(topic);
-        if (metadata.fetch().partitionsForTopic(topic) != null)
-            return 0;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:517]: Sorry for not noticing this earlier, but it seems that we now do two additional `time.milliseconds()` calls if `partitionsForTopic(topic) != null` compared to before. When producing small messages, these calls can be a bottleneck (see https://github.com/apache/kafka/pull/632 for example). Is there a way we can avoid this?

Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:517]: @ijuma Oops, thank you for pointing that out. I have updated the code to avoid the unnecessary `time.milliseconds()`. The code uses an AtomicLong to pass-by-reference (there is another similar usage in the same file), but I dont mind adding a class to return both cluster and wait time if that is preferred.

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:517]: Thanks. I prefer to avoid using `Atomic*` classes when we don't need atomicity guarantees. I introduced a `LongRef` class in the broker for this reason.

I think the right solution long-term is probably to introduce a class to complement `Time` that caches the last returned value so that one can choose when to refresh the value (for performance reasons). This was suggested some time ago in a different PR.

For this PR, as per your suggestion, I think using a private static class with two public fields is probably the way forward (since Java doesn't have tuples).

Comment (rajinisivaram) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:517]: @ijuma Thanks, done.

**Comment by ijuma** [comment]
Thanks @rajinisivaram, LGTM.
